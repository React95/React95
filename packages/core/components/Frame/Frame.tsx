import React, { forwardRef } from 'react';
import type {
  ComponentPropsWithRef,
  ElementType,
  ForwardedRef,
  Ref,
  ReactNode,
  RefAttributes,
} from 'react';
import { Sprinkles, sprinkles } from './Frame.css';
import cn from 'classnames';

type FixedForwardRef = <T, P = object>(
  render: (props: P, ref: Ref<T>) => ReactNode,
) => (props: P & RefAttributes<T>) => ReactNode;

export const fixedForwardRef = forwardRef as FixedForwardRef;

type DistributiveOmit<T, TOmitted extends PropertyKey> = T extends any
  ? Omit<T, TOmitted>
  : never;

export type FrameProps<TAs extends ElementType> = {
  as?: TAs;
} & DistributiveOmit<
  ComponentPropsWithRef<ElementType extends TAs ? 'div' : TAs>,
  'as'
> &
  Sprinkles;

const FrameComponent = <TAs extends ElementType = 'div'>(
  props: FrameProps<TAs>,
  ref: ForwardedRef<any>,
) => {
  const { as, children, ...rest } = props;
  const Component = as || 'div';
  const { className, style, otherProps } = sprinkles(rest);

  return (
    <Component
      {...otherProps}
      style={{ ...style, ...otherProps.style }}
      className={cn(className, otherProps.className)}
      ref={ref}
    >
      {children}
    </Component>
  );
};

export const Frame = fixedForwardRef(FrameComponent);
