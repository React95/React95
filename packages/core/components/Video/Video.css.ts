import { globalStyle, style } from '@vanilla-extract/css';
import { calc } from '@vanilla-extract/css-utils';
import { recipe } from '@vanilla-extract/recipes';

import { R95VideoNumbers } from '../shared/font-names';
import { contract } from '../themes/contract.css';

export const videoTag = recipe({
  base: {
    width: '100%',
    padding: contract.space[2],
  },
  variants: {
    visible: {
      true: {
        display: 'block',
      },
      false: {
        display: 'none',
      },
    },
  },
});

export const controls = style({
  display: 'flex',
  alignItems: 'center',
  paddingTop: contract.space[2],
  paddingBottom: contract.space[2],
});

export const countDownContainer = style({
  display: 'flex',
  padding: contract.space[6],
  marginBottom: contract.space[4],
  boxShadow: contract.shadows.in,
  backgroundColor: contract.colors.canvas,
  height: '50px',
  color: contract.colors.canvasText,
});

export const videoFont = style({
  fontFamily: R95VideoNumbers,
  textTransform: 'uppercase',
});

export const currentTime = style({
  marginTop: 'auto',
  fontSize: 22,
});

export const elapsedTime = style({
  height: contract.space[12],
});

export const divider = style({
  display: 'block',
  height: contract.space[1],
  borderTopStyle: 'solid',
  borderTopWidth: contract.space[1],
  borderTopColor: contract.colors.borderDark,
  borderBottomWidth: contract.space[1],
  borderBottomStyle: 'solid',
  borderBottomColor: contract.colors.borderLightest,
  marginBottom: contract.space[2],
});

export const controlBtn = style({
  display: 'inline-flex',
  justifyContent: 'center',
  alignItems: 'center',
  width: contract.space[20],
  height: contract.space[20],
  padding: contract.space[7],
  selectors: {
    '&:active, &:focus': {
      width: contract.space[20],
      height: contract.space[20],
      padding: contract.space[7],
    },
    '&:disabled': {
      padding: contract.space[4],
    },
  },
});

globalStyle(`${controlBtn} svg`, {
  fill: contract.colors.materialText,
});

globalStyle(`${controlBtn}:disabled svg`, {
  fill: contract.colors.borderDark,
  borderBottomWidth: contract.space[1],
  borderBottomStyle: 'solid',
  borderBottomColor: contract.colors.borderLightest,
  borderRightWidth: contract.space[1],
  borderRightStyle: 'solid',
  borderRightColor: contract.colors.borderLightest,
});

export const range = style({
  selectors: {
    '&::-webkit-slider-thumb': {
      height: contract.space[18],
      marginTop: calc.negate(contract.space[7]),
      width: contract.space[10],
    },
    "&[value='0']::-webkit-slider-thumb": {
      marginLeft: calc.negate(contract.space[2]),
    },
  },
});
