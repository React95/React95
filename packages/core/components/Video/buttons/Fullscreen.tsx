import React from 'react';
import type { HTMLAttributes } from 'react';

const Fullscreen = (props: HTMLAttributes<SVGElement>) => (
  <svg
    xmlns="http://www.w3.org/2000/svg"
    width="6"
    viewBox="0 0 24 24"
    height="6"
    aria-label="fullscreen"
    {...props}
  >
    <path
      d="M24 9h-2v-5h-7v-2h9v7zm-9 13v-2h7v-5h2v7h-9zm-15-7h2v5h7v2h-9v-7zm9-13v2h-7v5h-2v-7h9zm11 4h-16v12h16v-12z"
      style={{ width: 1000 }}
    />
  </svg>
);

export default Fullscreen;
