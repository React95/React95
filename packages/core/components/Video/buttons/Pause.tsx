import React from 'react';
import type { HTMLAttributes } from 'react';

const Pause = (props: HTMLAttributes<SVGElement>) => (
  <svg
    height="6"
    viewBox="0 0 424.236 424.236"
    width="6"
    xmlns="http://www.w3.org/2000/svg"
    aria-label="pause"
    {...props}
  >
    <path d="M256.471 2h176.765v424.236H256.471zM2 2h176.765v424.236H2z" />
  </svg>
);

export default Pause;
