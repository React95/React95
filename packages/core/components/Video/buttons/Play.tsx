import React from 'react';
import type { HTMLAttributes } from 'react';

const Play = (props: HTMLAttributes<SVGElement>) => (
  <svg
    height="6"
    viewBox="0 0 494.942 494.942"
    width="6"
    xmlns="http://www.w3.org/2000/svg"
    aria-label="play"
    {...props}
  >
    <path d="M35.353 0l424.236 247.471L35.353 494.942z" />
  </svg>
);

export default Play;
