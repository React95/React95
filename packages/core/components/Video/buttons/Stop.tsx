import React from 'react';
import type { HTMLAttributes } from 'react';

const Stop = (props: HTMLAttributes<SVGElement>) => (
  <svg
    xmlns="http://www.w3.org/2000/svg"
    width="6"
    height="6"
    aria-label="stop"
    {...props}
  >
    <path d="M0 0h306v306H0z" />
  </svg>
);

export default Stop;
