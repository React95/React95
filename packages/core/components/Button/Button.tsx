import React from 'react';
import type {
  ElementType,
  ReactElement,
  ElementRef,
  ForwardedRef,
  ButtonHTMLAttributes,
} from 'react';
import cn from 'classnames';

import { button } from './Button.css';
import { Frame, FrameProps, fixedForwardRef } from '../Frame/Frame';

export type ButtonProps<TAs extends ElementType> =
  ButtonHTMLAttributes<HTMLButtonElement> & FrameProps<TAs>;

export const Button = fixedForwardRef<HTMLButtonElement, ButtonProps<'button'>>(
  (props, ref) => {
    return (
      <Frame
        as="button"
        {...props}
        className={cn(button, props.className)}
        ref={ref}
      />
    );
  },
) as <TAs extends ElementType = 'button'>(
  props: ButtonProps<TAs> & { ref?: ForwardedRef<ElementRef<TAs>> },
) => ReactElement;
