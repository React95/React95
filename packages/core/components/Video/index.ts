import { Video } from './Video';

export default Video;
