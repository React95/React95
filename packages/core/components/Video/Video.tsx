import React, {
  useEffect,
  useState,
  useRef,
  forwardRef,
  useImperativeHandle,
} from 'react';
import type { FC, Ref, HTMLProps } from 'react';
import { Mplayer113, User4 } from '@react95/icons';

import * as styles from './Video.css';

import { Frame, FrameProps } from '../Frame/Frame';
import { Button } from '../Button/Button';
import { Range } from '../Range/Range';
import { TitleBar } from '../TitleBar/TitleBar';
import { Play, Pause, Stop, Fullscreen } from './buttons';
import cn from 'classnames';

type SourceProps = Pick<HTMLSourceElement, 'src'>;

const Source: FC<SourceProps> = ({ src }) => (
  <source src={src} type={`video/${src.substring(src.length - 3)}`} />
);

const PlayOrPause = ({ playing }: { playing: boolean }) =>
  playing ? <Pause /> : <Play />;

const arrayFy = (str: string | string[]) => ([] as string[]).concat(str);

function updateProgressBar(
  player: HTMLVideoElement,
  updateProgress: (value: number) => void,
) {
  const percentage = Math.floor((100 / player.duration) * player.currentTime);

  updateProgress(percentage);
}

function parseCurrentTime(secs: number): string {
  if (!secs) {
    return '00:00';
  }

  const sec = parseInt(secs.toString(), 10);
  const hours = Math.floor(sec / 3600);
  const minutes = Math.floor(sec / 60) % 60;
  const seconds = sec % 60;

  return [hours, minutes, seconds]
    .map(v => (v < 10 ? `0${v}` : v))
    .filter((v, i) => v !== '00' || i > 0)
    .join(':');
}

export type VideoProps = {
  name?: string;
  src: string;
  videoProps?: HTMLProps<HTMLVideoElement>;
} & FrameProps<'div'>;

export type VideoRefs = {
  video: Ref<HTMLVideoElement>;
  progress: Ref<HTMLInputElement>;
  wrapper: Ref<HTMLDivElement>;
  playpause: Ref<HTMLButtonElement>;
  stop: Ref<HTMLButtonElement>;
  fullScreen: Ref<HTMLButtonElement>;
};

const VideoRenderer = (
  { name, src, videoProps, ...props }: VideoProps,
  ref: Ref<VideoRefs>,
) => {
  const [playing, setPlaying] = useState(false);
  const [loadeddata, setLoadeddata] = useState(false);
  const [progress, setProgress] = useState(0);

  const player = useRef<HTMLVideoElement>(null);
  const progressRef = useRef<HTMLInputElement>(null);
  const wrapperRef = useRef<HTMLDivElement>(null);
  const playPauseRef = useRef<HTMLButtonElement>(null);
  const stopRef = useRef<HTMLButtonElement>(null);
  const fullScreenRef = useRef<HTMLButtonElement>(null);

  const paths = arrayFy(src);
  const [pathname] = paths;

  const normalizedName = name || pathname.replace(/^.*[\\/]/, '');
  const title = `${normalizedName}${!loadeddata ? ' (Opening)' : ''}`;

  useImperativeHandle(ref, () => ({
    get video() {
      return player;
    },
    get progress() {
      return progressRef;
    },
    get wrapper() {
      return wrapperRef;
    },
    get playpause() {
      return playPauseRef;
    },
    get stop() {
      return stopRef;
    },
    get fullScreen() {
      return fullScreenRef;
    },
  }));

  useEffect(() => {
    player.current?.addEventListener(
      'ended',
      () => {
        setPlaying(false);
        setProgress(0);
      },
      false,
    );

    player.current?.addEventListener(
      'timeupdate',
      () => {
        if (player.current) {
          updateProgressBar(player.current, setProgress);
        }
      },
      false,
    );

    player.current?.addEventListener(
      'loadeddata',
      () => {
        setLoadeddata(true);
      },
      false,
    );

    player.current?.addEventListener(
      'playing',
      () => {
        setPlaying(true);
      },
      false,
    );
  }, [player.current]);

  return (
    <Frame
      {...props}
      padding="$2"
      boxShadow="$out"
      backgroundColor="$material"
      ref={wrapperRef}
    >
      <TitleBar icon={<Mplayer113 variant="16x16_4" />} title={title} />
      <video
        className={styles.videoTag({ visible: loadeddata })}
        {...videoProps}
        ref={player}
      >
        {paths.map(s => (
          <Source key={s} src={s} />
        ))}
      </video>
      {loadeddata && <span className={styles.divider} />}
      <Frame maxWidth="250px" mx="auto" mb="$4">
        <div className={styles.countDownContainer}>
          <Frame display="flex" flexDirection="column" w="40%">
            <div
              className={styles.videoFont}
              style={{
                marginTop: 'auto',
              }}
            >
              {player.current && parseCurrentTime(player.current.duration)}
            </div>

            <div className={styles.videoFont} style={{ height: 12 }}>
              {!loadeddata && 'Openning'}
            </div>
          </Frame>
          <Frame display="flex" flexDirection="column" w="40%">
            <div className={cn(styles.videoFont, styles.currentTime)}>
              {player.current && parseCurrentTime(player.current.currentTime)}
            </div>

            <div className={cn(styles.videoFont, styles.elapsedTime)}>time</div>
          </Frame>
        </div>
        <div className={styles.controls}>
          <Button
            as="button"
            className={styles.controlBtn}
            disabled={!loadeddata}
            onClick={() => {
              if (!playing) {
                player.current?.play();
              } else {
                player.current?.pause();
              }
              setPlaying(!playing);
            }}
            ref={playPauseRef}
          >
            {loadeddata ? (
              <PlayOrPause playing={playing} />
            ) : (
              <User4
                style={{ borderRight: 'none', borderBottom: 'none' }}
                variant="32x32_4"
              />
            )}
          </Button>
          <Button
            className={styles.controlBtn}
            disabled={!loadeddata}
            onClick={() => {
              if (player.current) {
                player.current.pause();
                player.current.currentTime = 0;
              }

              setPlaying(false);
            }}
            ref={stopRef}
          >
            <Stop />
          </Button>
          <Button
            className={styles.controlBtn}
            disabled={!loadeddata}
            onClick={() => {
              player?.current?.requestFullscreen();
            }}
            ref={fullScreenRef}
          >
            <Fullscreen />
          </Button>

          <Range
            className={styles.range}
            ref={progressRef}
            min="0"
            max="100"
            step="1"
            value={progress}
            style={{
              width: '70%',
              marginLeft: 20,
            }}
            onChange={({ target }) => {
              const { current: video } = player;

              if (video) {
                const value = parseInt(target.value);
                const percent = value / 100;

                video.currentTime = percent * video.duration;

                setProgress(value);
              }
            }}
          />
        </div>
      </Frame>
    </Frame>
  );
};

export const Video = forwardRef<VideoRefs, VideoProps>(VideoRenderer);
