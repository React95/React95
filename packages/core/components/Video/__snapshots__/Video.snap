// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Video /> > Snapshot > should match snapshot 1`] = `
<div>
  <div
    class="Frame_padding__iuzg892r Frame_boxShadow__iuzg895z Frame_backgroundColor__iuzg893p"
    style="--padding__iuzg892q: var(--r95-space-2); --boxShadow__iuzg895y: var(--r95-shadow-out); --backgroundColor__iuzg893o: var(--r95-color-material);"
  >
    <div
      class="TitleBar_titleBarBackground__1kpzw682 TitleBar_titleBarBackground_active_true__1kpzw683"
    >
      <svg test-file-stub
        variant="16x16_4"
      />
      <div
        class="TitleBar_title__1kpzw685"
      >
        some_video.mp4 (Opening)
      </div>
    </div>
    <video
      class="Video_videoTag__13gnpv00 Video_videoTag_visible_false__13gnpv02"
    >
      <source
        src="foo/bar/some_video.mp4"
        type="video/mp4"
      />
    </video>
    <div
      class="Frame_marginBottom__iuzg892d Frame_marginLeft__iuzg892l Frame_marginRight__iuzg892n Frame_maxWidth__iuzg891v"
      style="--marginBottom__iuzg892c: var(--r95-space-4); --marginLeft__iuzg892k: auto; --marginRight__iuzg892m: auto; --maxWidth__iuzg891u: 250px;"
    >
      <div
        class="Video_countDownContainer__13gnpv04"
      >
        <div
          class="Frame_width__iuzg891t Frame_display__iuzg891 Frame_flexDirection__iuzg8915"
          style="--width__iuzg891s: 40%; --display__iuzg890: flex; --flexDirection__iuzg8914: column;"
        >
          <div
            class="Video_videoFont__13gnpv05"
            style="margin-top: auto;"
          />
          <div
            class="Video_videoFont__13gnpv05"
            style="height: 12px;"
          >
            Openning
          </div>
        </div>
        <div
          class="Frame_width__iuzg891t Frame_display__iuzg891 Frame_flexDirection__iuzg8915"
          style="--width__iuzg891s: 40%; --display__iuzg890: flex; --flexDirection__iuzg8914: column;"
        >
          <div
            class="Video_videoFont__13gnpv05 Video_currentTime__13gnpv06"
          />
          <div
            class="Video_videoFont__13gnpv05 Video_elapsedTime__13gnpv07"
          >
            time
          </div>
        </div>
      </div>
      <div
        class="Video_controls__13gnpv03"
      >
        <button
          class="Button_button__1bzq74l0 Video_controlBtn__13gnpv09"
          disabled=""
        >
          <svg test-file-stub
            style="border-right-width: initial; border-right-style: none; border-right-color: none; border-bottom-width: initial; border-bottom-style: none; border-bottom-color: none;"
            variant="32x32_4"
          />
        </button>
        <button
          class="Button_button__1bzq74l0 Video_controlBtn__13gnpv09"
          disabled=""
        >
          <svg
            aria-label="stop"
            height="6"
            width="6"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h306v306H0z"
            />
          </svg>
        </button>
        <button
          class="Button_button__1bzq74l0 Video_controlBtn__13gnpv09"
          disabled=""
        >
          <svg
            aria-label="fullscreen"
            height="6"
            viewBox="0 0 24 24"
            width="6"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M24 9h-2v-5h-7v-2h9v7zm-9 13v-2h7v-5h2v7h-9zm-15-7h2v5h7v2h-9v-7zm9-13v2h-7v5h-2v-7h9zm11 4h-16v12h16v-12z"
              style="width: 1000px;"
            />
          </svg>
        </button>
        <input
          class="Range_range__1i4w9z20 Video_range__13gnpv0a"
          max="100"
          min="0"
          step="1"
          style="width: 70%; margin-left: 20px;"
          type="range"
          value="0"
        />
      </div>
    </div>
  </div>
</div>
`;
